proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,271 @@
+  the source program
+* the cost-centre profiler (accessible from Haskell with
+  ``GHC.Stack.CCS.getCurrentCCS``) can provide cost-center stacks
+* DWARF debug information in conjunction with GHC's `built-in stack unwinder
+  <https://www.haskell.org/ghc/blog/20200405-dwarf-3.html>`_ can be used
+  to provide terse (but nevertheless useful) backtraces with no runtime
+  overhead in the non-failing case (although backtrace collection is quite
+  slow)
+
+All of these backtrace options have their time and place, offering a range of
+levels of detail, executable size, and runtime overhead. GHC, being a compiler,
+shouldn't be in the business of dictating which of these mechanisms should be
+used to report exception backtraces.  Consequently, our ``Backtrace`` type is
+designed to capture them all: ::
+
+    -- | An exception backtrace.
+    data Backtrace
+        = CostCenterBacktrace (Ptr GHC.Stack.CCS.CostCentreStack)
+          -- ^ a cost center profiler backtrace
+        | HasCallStackBacktrace GHC.Stack.CallStack
+          -- ^ a stack from 'GHC.Stack.HasCallStack'
+        | ExecutionBacktrace [GHC.ExecutionStack.Location]
+          -- ^ a stack unwinding (e.g. DWARF) backtrace
+
+    instance Show Backtrace
+
+With such a type we can easily write a variant of ``throwIO`` that, for
+instance, attaches a ``HasCallStack`` backtrace: ::
+
+    -- | Throws an exception with a 'HasCallStack' backtrace.
+    throwIOWithCallStack :: (Exception e, HasCallStack) => e -> IO a
+    throwIOWithCallStack exc = throw callStack exc
+
+We propose that ``GHC.Exception`` provide a family of these functions for
+the ``HasCallStack``, cost-center stack, and execution stack cases.
+
+Making backtraces ubiquitous
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+While these functions are useful building blocks, they do not
+address the most common debugging scenario: searching for an exception
+thrown by a third-party library (which likely calls the old ``throw`` and
+consequently would not produce a backtrace). For this we propose a pragmatic,
+stateful approach to allow the user to select which mechanism should be used
+for backtrace collection in ``mkSomeExceptionWithLocation``: ::
+
+    module GHC.Exception.Backtrace where
+
+    -- | Which kind of backtrace to collect when an exception is thrown.
+    data BacktraceMechanism
+      = NoBacktrace
+      | CollectCostCenterBacktrace
+      | HasCallStackBacktrace
+      | ExecutionStackBacktrace
+
+    -- | Set the global backtrace mechanism.
+    -- (this state would be represented internally as a simple global IORef.
+    setGlobalBacktraceMechanism :: BacktraceMechanism -> IO ()
+
+The ``collectBacktrace`` primitive referenced in the previous section then
+simply dispatches to the appropriate backtrace collection scheme as determined
+by the currently selected ``BacktraceMechanism``.
+
+Examples
+--------
+
+User programs would typically call ``setGlobalBacktraceMechanism`` during
```

Comment:
I'm really nitpicking here. But I don't think “global backtrace mechanism” characterises this option so well. Rather, I would think of it more of a “default backtrack mechanism”.

Reply:
That is a fair point.

Reply:
@aspiwack , @bgamari ,

Is this agreed?

Should I use `setDefaultBacktraceMechanisms` as name for this function? (I've added a plural "s" to the name, because one can set a list of mechanisms.)

Reply:
It's what the proposal currently says. So it's probably agreed :slightly_smiling_face: